lib/fluent/plugin/out_secondary_file.rb

```diff
@@ -117,15 +116,45 @@ def has_time_format?(str)
 
     def generate_path(path_without_suffix)
       if @append
-        "#{path_without_suffix}#{@suffix}"
-      else
+        path = "#{path_without_suffix}#{@suffix}"
+        lock_if_needed(path) do
+          yield path
+        end
+        return path
+      end
+
+      begin
         i = 0
         loop do
           path = "#{path_without_suffix}.#{i}#{@suffix}"
-          return path unless File.exist?(path)
+          break unless File.exist?(path)
           i += 1
         end
+        lock_if_needed(path) do
+          # If multiple processes select the same path and another process
+          # entered this locking block first, the file should already exist
+          # and this process should retry to find new path.
+          raise FileAlreadyExist if File.exist?(path)
+          yield path
+        end
```

Comment:
I just noticed that this logic perhaps must consider working on multiple threads, not only multiple processes.
I'll do additional tests and make the fix if necessary.

Reply:
Test for fe65b4f37730e718ec23528142429f42d0961515.

Use the same way and the same environment as

* #4080#issuecomment-1455695378

Result: 

| workers | flush_thread_count |   percentage  |
|---------|--------------------|---------------|
|       1 |                 16 | 45.5% (10/22) |
|       3 |                  1 | 0.0% (0/51)   |

So the current fix is not enough.